include/seqan3/range/view/take_until.hpp

```diff
@@ -170,6 +167,135 @@ class view_take_until : public ranges::view_interface<view_take_until<urng_t, fu
+        ranges::semiregular_t<fun_ref_t> fun;
+
+        //!\brief The sentinel type is identical to that of the underlying range.
+        using sentinel_type = std::ranges::sentinel_t<rng_t>;
+
+        //!\brief Whether this iterator has reached the end (cache is only used on pure input ranges).
+        sentinel_type stored_end;
+
+        //!\brief Whether the end was reached by evaluating the functor.
+        bool at_end_gracefully = false;
+
+    public:
+        /*!\name Constructors, destructor and assignment
+         * \{
+         */
+        iterator_type_consume_input()                                                         = default; //!< Defaulted.
+        constexpr iterator_type_consume_input(iterator_type_consume_input const & rhs)        = default; //!< Defaulted.
+        constexpr iterator_type_consume_input(iterator_type_consume_input && rhs)             = default; //!< Defaulted.
+        //!\brief Defaulted.
+        constexpr iterator_type_consume_input & operator=(iterator_type_consume_input const & rhs) = default;
+        constexpr iterator_type_consume_input & operator=(iterator_type_consume_input && rhs) = default; //!< Defaulted.
+        ~iterator_type_consume_input() = default;
+
+        //!\brief Constructor that delegates to the CRTP layer.
+        iterator_type_consume_input(base_base_t const & it) :
+            base_t{it}
+        {}
+
+        //!\brief Constructor that delegates to the CRTP layer and initialises the callable.
+        iterator_type_consume_input(base_base_t it, fun_ref_t _fun, sentinel_type sen) :
+            base_t{std::move(it)}, fun{_fun}, stored_end{std::move(sen)}
+        {}
+        //!\}
+
+        /*!\name Associated types
+         * \brief All are derived from the base_base_t.
+         * \{
+         */
+        using difference_type       = typename std::iterator_traits<base_base_t>::difference_type;  //!< From base.
+        using value_type            = typename std::iterator_traits<base_base_t>::value_type;       //!< From base.
+        using reference             = typename std::iterator_traits<base_base_t>::reference;        //!< From base.
+        using pointer               = typename std::iterator_traits<base_base_t>::pointer;          //!< From base.
+        using iterator_category     = std::input_iterator_tag;                                      //!< Always input.
+        //!\}
+
+        /*!\name Arithmetic operators
+         * \brief seqan3::detail::inherited_iterator_base operators are used unless specialised here.
+         * \{
+         */
+        //!\brief Override pre-increment to implement consuming behaviour.
+        iterator_type_consume_input & operator++() noexcept(!or_throw)
+        {
+            ++static_cast<base_base_t &>(*this);
+
+            if ((static_cast<base_base_t &>(*this) != stored_end) && fun(**this))
+            {
+                at_end_gracefully = true;
+                ++(*this); // call this function recursively
```

Comment:
Why are you calling it recursively ? If it is to fully consume `\r\n` does it also consume `\n\n`? This could be a problem for multiline parsing as I would not expect to consume more than one newline.

Reply:
hm this really unintuitive because our take_until stops at the FIRST occurrence right?

Additionally I see the following problems for IO:
1) there might be formats where empty (e.g. tab separated) fields are allowed (luckily it is not in SAM), what about BLAST?
2) we automatically allow wrongly formatted files which include an extra newline

Is this recursion only because of the possibility of (`\r\n` as new line)?

Reply:
not resolved?

Reply:
> This does not change the existing take_until

I know, but it is called similar but behaves differently, which is unintuitive.

What about my other concerns/questions?

Reply:
> I know, but it is called similar but behaves differently, which is unintuitive.

How would you call it? The returned range is identical to that of take_until...

> What about my other concerns/questions?

> there might be formats where empty (e.g. tab separated) fields are allowed (luckily it is not in SAM), what about BLAST?

How is this related to my changes? I am not changing either of the formats...

> we automatically allow wrongly formatted files which include an extra newline

For the last line we currently accept newline or no-newline in any case. Accepting two newlines is not a breaking change since we agreed that we don't validate formats by default anyway, we just try to extract the information we need.

Reply:
> How would you call it? The returned range is identical to that of take_until...

I would call it the same but change the behaviour to that it does not consume all character matching `fun` but just the first. The consuming of `\r\n` should be part of `take_line` not `take_until_and_comsume`.

> How is this related to my changes? I am not changing either of the formats...

well I thought you added this functionality specifically for the use in our formats

> For the last line we currently accept newline or no-newline in any case. Accepting two newlines is not a breaking change since we agreed that we don't validate formats by default anyway, we just try to extract the information we need.

we will also allow

```
> seq1
ACTACTAG


ACTAGC
AC
```

But ok you are right that we do not validate formats anyway.

Reply:
> I would call it the same but change the behaviour to that it does not consume all character matching fun but just the first.

But I added this functionality explicitly so that we can use it in I/O and don't have to call detail::consume so often. It's useful for skipping multiple whitespaces among other things.

> well I thought you added this functionality specifically for the use in our formats

Yes, and it works for those situations, or not? Skipping a single character is easy to do, you can just cal `++it;` , but skipping multiple ones involves another `detail::consume(foo | view::take_until(...))` which was what I was trying to avoid.

> we will also allow

That's already valid, because we take_until(is_char<'>'>) and filter whitespaces...